pkg/queue/queue.go

```diff
@@ -135,80 +136,110 @@ func (q *RequestQueue) ReleaseRequests(items []Request) {
 }
 
 // DequeueMany consumes multiple items for a single tenant from the queue.
-// It returns maxItems and waits maxWait if no requests for this tenant are enqueued.
+// It blocks the execution until it dequeues at least 1 request and continue reading
+// until it reaches `maxItems` requests or if no requests for this tenant are enqueued.
 // The caller is responsible for returning the dequeued requests back to the
 // pool by calling ReleaseRequests(items).
-func (q *RequestQueue) DequeueMany(ctx context.Context, last QueueIndex, consumerID string, maxItems int, maxWait time.Duration) ([]Request, QueueIndex, error) {
-	// create a context for dequeuing with a max time we want to wait to fulfill the desired maxItems
-
-	dequeueCtx, cancel := context.WithTimeout(ctx, maxWait)
-	defer cancel()
-
-	var idx QueueIndex
-
+func (q *RequestQueue) DequeueMany(ctx context.Context, idx QueueIndex, consumerID string, maxItems int) ([]Request, QueueIndex, error) {
 	items := q.pool.Get(maxItems)
+	lastQueueName := anyQueueName
 	for {
-		item, newIdx, err := q.Dequeue(dequeueCtx, last, consumerID)
+		item, newIdx, newQueueName, tenantQueueEmpty, err := q.dequeue(ctx, idx, lastQueueName, consumerID)
 		if err != nil {
-			if err == context.DeadlineExceeded {
+			// the consumer must receive the items if tenants queue is removed,
+			// even if it has collected less than `maxItems` requests.
+			if errors.Is(err, ErrQueueWasRemoved) {
 				err = nil
 			}
-			return items, idx, err
+			return items, newIdx, err
 		}
+		lastQueueName = newQueueName
 		items = append(items, item)
-		idx = newIdx
-		if len(items) == maxItems {
-			return items, idx, nil
+		idx = newIdx.ReuseLastIndex()
+		if len(items) == maxItems || tenantQueueEmpty {
+			return items, newIdx, nil
 		}
 	}
 }
 
+const anyQueueName = ""
```

Comment:
nit: move them to the other const at the top of the file. also, maybe name it `anyQueue`
The queue does not really have a concept of "named queues". The names are the tenant ids by coincident.

Reply:
done